### PR TITLE
Generate recon config from database and schema

### PR DESCRIPTION
**Pull request by vijaypavann-db**

### CODE REVIEW

File: labs.yml [hunk 1]
```diff
@@ -46,3 +46,12 @@ commands:
       - name: report
         default: all
         description: Report type
+  - name: setup-recon-secrets
```
Comment (sundarshankar89) [inline labs.yml:49]:
```suggestion
  - name: configure-secrets
```
This makes it more generic and helps us expand this command if needed

File: src/databricks/labs/remorph/cli.py [hunk 2]
```diff
@@ -20,13 +24,13 @@ def raise_validation_exception(msg: str) -> Exception:
...
 @remorph.command
 def transpile(
-    w: WorkspaceClient,
-    source: str,
-    input_sql: str,
-    output_folder: str,
-    skip_validation: str,
-    catalog_name: str,
-    schema_name: str,
+        w: WorkspaceClient,
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/cli.py:27]: make fmt should fix these kind of differences

File: src/databricks/labs/remorph/helpers/db_workspace_utils.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+logger = get_logger(__file__)
+
+
+class DBWorkspaceClient:
+    def __init__(self, ws: WorkspaceClient, prompts: Prompts):
+        self._ws = ws
+        self._prompts = prompts
+
+    def scope_exists(self, scope_name: str) -> bool:
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/helpers/db_workspace_utils.py:14]: Make this as private method

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py [hunk 4]
```diff
@@ -52,7 +65,11 @@ def _get_jdbc_reader_options(options: JdbcReaderOptions):
 
     def _get_secrets(self, key_name: str):
         key = self.engine + '_' + key_name
-        return self.ws.secrets.get_secret(self.scope, key)
+        dbutils = self.ws.dbutils
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/connectors/data_source.py:68]: please convert bytes to actual string don't use dbutils directly.  using a decoder

File: src/databricks/labs/remorph/reconcile/connectors/snowflake.py [hunk 5]
```diff
@@ -1,28 +1,43 @@
...
     @property
     def get_jdbc_url(self) -> str:
-        return (
+        jdbc_url = (
             f"jdbc:{SourceType.SNOWFLAKE.value}://{self._get_secrets('account')}.snowflakecomputing.com"
             f"/?user={self._get_secrets('sfUser')}&password={self._get_secrets('sfPassword')}"
             f"&db={self._get_secrets('sfDatabase')}&schema={self._get_secrets('sfSchema')}"
             f"&warehouse={self._get_secrets('sfWarehouse')}&role={self._get_secrets('sfRole')}"
         )
+        logger.debug(f"JDBC URL: {jdbc_url}")
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/connectors/snowflake.py:33]: property shouldn't have loggers, logger should be implemented outside where the property is called.

File: src/databricks/labs/remorph/helpers/recon_config_utils.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        sf_url = self._prompts.question("Enter Snowflake URL")
+        account = self._prompts.question("Enter Account Name")
+        sf_user = self._prompts.question("Enter User")
+        sf_password = self._prompts.question("Enter Password")
+        sf_db = self._prompts.question("Enter Database")
+        sf_schema = self._prompts.question("Enter Schema")
+        sf_warehouse = self._prompts.question("Enter Snowflake Warehouse")
+        sf_role = self._prompts.question("Enter Role", default=" ")
+
+        sf_conn_details = [
```
Comment (vil1) [inline src/databricks/labs/remorph/helpers/recon_config_utils.py:49]: Why not defining this directly as a dict?

File: src/databricks/labs/remorph/helpers/recon_config_utils.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+    def prompt_oracle_connection_details(self) -> tuple[str, dict[str, str]]:
+        logger.info(f"Please answer a couple of questions to configure `{SourceType.ORACLE.value}` Connection profile")
+        user = self._prompts.question("Enter User")
+        password = self._prompts.question("Enter Password")
+        host = self._prompts.question("Enter host")
+        port = self._prompts.question("Enter port")
+        database = self._prompts.question("Enter database/SID")
+
+        oracle_conn_details = [
```
Comment (vil1) [inline src/databricks/labs/remorph/helpers/recon_config_utils.py:70]: Same comment as above

File: src/databricks/labs/remorph/helpers/recon_config_utils.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+
+    def prompt_snowflake_connection_details(self) -> tuple[str, dict[str, str]]:
+        logger.info(
+            f"Please answer a couple of questions to configure `{SourceType.SNOWFLAKE.value}` Connection profile"
+        )
+
+        sf_url = self._prompts.question("Enter Snowflake URL")
+        account = self._prompts.question("Enter Account Name")
+        sf_user = self._prompts.question("Enter User")
+        sf_password = self._prompts.question("Enter Password")
```
Comment (vil1) [inline src/databricks/labs/remorph/helpers/recon_config_utils.py:43]: Does the `question` method know that it should not print the user's input when asking for a password or other kind of sensitive information?

File: labs.yml (hunk 1 above)
Comment (vijaypavann-db) [reply labs.yml:49]: Done

File: src/databricks/labs/remorph/helpers/recon_config_utils.py (hunk 6 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/helpers/recon_config_utils.py:49]: Done

File: src/databricks/labs/remorph/helpers/recon_config_utils.py (hunk 8 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/helpers/recon_config_utils.py:43]: I don't think so, it shows the sensitive information on screen. But after that, we're storing that information in configured `Databricks Secrets scope`. 
In `databricks cli` `Secrets` they have the provision to read from `json` /  `bytes value` / `string`. We will add this provision here [if needed]. 

> https://docs.databricks.com/en/dev-tools/cli/commands.html 

```
Usage:
  databricks secrets put-secret SCOPE KEY [flags]

Flags:
      --bytes-value string    If specified, value will be stored as bytes.
  -h, --help                  help for put-secret
      --json JSON             either inline JSON string or @path/to/file.json with request body (default JSON (0 bytes))
      --string-value string   If specified, note that the value will be stored in UTF-8 (MB4) form.
```

image

Comment (vil1) [reply src/databricks/labs/remorph/helpers/recon_config_utils.py:47]: The possibility of having sensitive information displayed on screen feels like a security liability, but I let @nfx decide whether it should prevent this PR from being merged as is.

**Comment by sundarshankar89** [comment]
@vijaypavann-db you can probably close this pr

File: src/databricks/labs/remorph/helpers/recon_config_utils.py (hunk 8 above)
Comment (nfx) [reply src/databricks/labs/remorph/helpers/recon_config_utils.py:49]: well... eventually we should switch to displaying the HTML page on localhost webserver, but that would come later.

**Review by nfx — CHANGES_REQUESTED** [review]
this PR is too big. split it to more granular PRs.

**Comment by vijaypavann-db** [comment]
This PR is split into three 3 PRs:
* [Generate Recon Config](https://github.com/databrickslabs/remorph/pull/249)
* [Validate Recon Config](https://github.com/databrickslabs/remorph/pull/250)
* [Configure Secrets](https://github.com/databrickslabs/remorph/pull/253)

Please review the associated PR for functionality.

**Pull request closed without merging**